src/journal/views.py

```diff
@@ -1392,36 +1392,73 @@ def search(request):
     """
     articles = []
     search_term = None
+    keyword = None
+    if request.POST and 'clear' in request.POST:
+        return logic.unset_search_GET_variables(request)
+        # return logic.unset_search_session_variables(request)
 
-    if request.POST:
-        search_term = request.POST.get('search')
-        request.session['article_search'] = search_term
-        return redirect(reverse('search'))
+    search_term, keyword, sort, redir = logic.handle_search_controls(request)
 
-    if request.session.get('article_search'):
-        search_term = request.session.get('article_search')
+    if redir:
+        return redir
+    #if no order, set order to title (alphabetical). this happens when receiveing a GET request for keyword.
+    if request.GET and not sort:
+        sort='title'
 
+    if search_term:
+        # return two lists of pks based on search term.
         article_search = submission_models.Article.objects.filter(
             (Q(title__icontains=search_term) |
-             Q(subtitle__icontains=search_term)) &
-            Q(journal=request.journal)
+             Q(keywords__word__icontains=search_term) |
+             Q(subtitle__icontains=search_term)), 
+            journal=request.journal, 
+            stage=submission_models.STAGE_PUBLISHED,
+            date_published__lte=timezone.now()
         )
-        article_search = [article for article in article_search]
+        article_search_unsorted = [article.pk for article in article_search]
 
-        author_search = search_term.split(' ')
         from_author = submission_models.FrozenAuthor.objects.filter(
-            (Q(first_name__in=author_search) |
-             Q(last_name__in=author_search)) &
-            Q(article__journal=request.journal)
+            (Q(first_name__icontains=search_term) |
+             Q(last_name__icontains=search_term)),
+            article__journal=request.journal, 
+            article__stage=submission_models.STAGE_PUBLISHED, 
+            article__date_published__lte=timezone.now()
         )
-
-        articles_from_author = [author.article for author in from_author]
-        articles = set(article_search + articles_from_author)
+        articles_from_author_unsorted = [author.article.pk for author in from_author]
+        # merge the two lists into a set which will clear out duplicates
+        article_list_unsorted = set(article_search_unsorted + articles_from_author_unsorted)
+        # do a search on the set of pks and apply the order_by while turning back into a list of article objects.
+        articles = list(submission_models.Article.objects.filter(pk__in=article_list_unsorted).order_by(sort))
```

Comment:
Since you are combining `search_term` and `author_search`, have you considered doing the filtering/sorting on the query instead of using intermediate python constructs? It would save us the extra database roundtrip as well as save us quite a bit of memory for large result sets. (This view will be very demanding on some of our journals).

Reply:
This was done on my sqlite database on my local which has relatively few articles.
it returned one article named `foo bur` and one article published by `Richard Burkhardt`
```
search_term = 'bur'
author_search = search_term.split(" ")
j = Journal.objects.get(pk=1)
a = submission_models.Article.objects.filter(
(Q(title__icontains=search_term) |
Q(keywords__word__icontains=search_term) |
Q(subtitle__icontains=search_term)) |
(Q(frozenauthor__first_name__icontains=search_term) |
Q(frozenauthor__last_name__icontains=search_term))
|
(Q(frozenauthor__first_name__in=author_search) |
Q(frozenauthor__last_name__in=author_search)), 
journal=j, 
stage=submission_models.STAGE_PUBLISHED,
date_published__lte=timezone.now()
).distinct()
>>> connection.queries
[{'sql': 'SELECT "journal_journal"."id", "journal_journal"."domain", "journal_journal"."is_secure", "journal_journal"."code", "journal_journal"."current_issue_id", "journal_journal"."carousel_id", "journal_journal"."thumbnail_image_id", "journal_journal"."press_image_override_id", "journal_journal"."default_cover_image", "journal_journal"."default_large_image", "journal_journal"."header_image", "journal_journal"."favicon", "journal_journal"."description", "journal_journal"."contact_info", "journal_journal"."disable_metrics_display", "journal_journal"."disable_article_images", "journal_journal"."enable_correspondence_authors", "journal_journal"."disable_html_downloads", "journal_journal"."full_width_navbar", "journal_journal"."is_remote", "journal_journal"."is_conference", "journal_journal"."remote_submit_url", "journal_journal"."remote_view_url", "journal_journal"."view_pdf_button", "journal_journal"."nav_home", "journal_journal"."nav_news", "journal_journal"."nav_articles", "journal_journal"."nav_issues", "journal_journal"."nav_collections", "journal_journal"."nav_contact", "journal_journal"."nav_start", "journal_journal"."nav_review", "journal_journal"."nav_sub", "journal_journal"."has_xslt", "journal_journal"."hide_from_press", "journal_journal"."sequence" FROM "journal_journal" WHERE "journal_journal"."id" = 1', 
'time': '0.009'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%rich%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:08:33.017880\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.002'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%man%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:08:52.257919\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21',
'time': '0.001'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%rich man%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%rich man%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:24:13.909479\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.002'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%man%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%man%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:24:48.261761\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21',
'time': '0.001'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%bur%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%bur%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:24:59.573736\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.001'}, 
{'sql': 'SELECT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%rich%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%rich%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:25:09.240307\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.001'}, 
{'sql': 'SELECT DISTINCT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%rich%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%rich%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:26:09.195157\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.002'}, 
{'sql': 'SELECT DISTINCT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_article_authors" ON ("submission_article"."id" = "submission_article_authors"."article_id") LEFT OUTER JOIN "core_account" ON ("submission_article_authors"."account_id" = "core_account"."id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%rich man%\' ESCAPE \'\\\' OR "core_account"."first_name" LIKE \'%rich man%\' ESCAPE \'\\\' OR "core_account"."last_name" LIKE \'%rich man%\' ESCAPE \'\\\' OR "core_account"."first_name" IN (\'rich\', \'man\') OR "core_account"."last_name" IN (\'rich\', \'man\')) AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:27:20.602862\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.002'}, 
{'sql': 'SELECT "submission_frozenauthor"."id", "submission_frozenauthor"."article_id", "submission_frozenauthor"."author_id", "submission_frozenauthor"."first_name", "submission_frozenauthor"."middle_name", "submission_frozenauthor"."last_name", "submission_frozenauthor"."institution", "submission_frozenauthor"."department", "submission_frozenauthor"."country_id", "submission_frozenauthor"."order" FROM "submission_frozenauthor" INNER JOIN "submission_article" ON ("submission_frozenauthor"."article_id" = "submission_article"."id") WHERE (("submission_frozenauthor"."first_name" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."last_name" LIKE \'%rich man%\' ESCAPE \'\\\') AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:27:29.632774\') ORDER BY "submission_frozenauthor"."order" ASC LIMIT 21', 
'time': '0.001'}, 
{'sql': 'SELECT "submission_frozenauthor"."id", "submission_frozenauthor"."article_id", "submission_frozenauthor"."author_id", "submission_frozenauthor"."first_name", "submission_frozenauthor"."middle_name", "submission_frozenauthor"."last_name", "submission_frozenauthor"."institution", "submission_frozenauthor"."department", "submission_frozenauthor"."country_id", "submission_frozenauthor"."order" FROM "submission_frozenauthor" INNER JOIN "submission_article" ON ("submission_frozenauthor"."article_id" = "submission_article"."id") WHERE (("submission_frozenauthor"."first_name" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."last_name" LIKE \'%rich man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" IN (\'rich\', \'man\') OR "submission_frozenauthor"."last_name" IN (\'rich\', \'man\')) AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:27:38.826195\') ORDER BY "submission_frozenauthor"."order" ASC LIMIT 21', 
'time': '0.001'}, 
{'sql': 'SELECT "submission_frozenauthor"."id", "submission_frozenauthor"."article_id", "submission_frozenauthor"."author_id", "submission_frozenauthor"."first_name", "submission_frozenauthor"."middle_name", "submission_frozenauthor"."last_name", "submission_frozenauthor"."institution", "submission_frozenauthor"."department", "submission_frozenauthor"."country_id", "submission_frozenauthor"."order" FROM "submission_frozenauthor" INNER JOIN "submission_article" ON ("submission_frozenauthor"."article_id" = "submission_article"."id") WHERE (("submission_frozenauthor"."first_name" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_frozenauthor"."last_name" LIKE \'%rich%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" IN (\'rich\') OR "submission_frozenauthor"."last_name" IN (\'rich\')) AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:28:05.735580\') ORDER BY "submission_frozenauthor"."order" ASC LIMIT 21', 
'time': '0.001'}, 
{'sql': 'SELECT DISTINCT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_frozenauthor" ON ("submission_article"."id" = "submission_frozenauthor"."article_id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."last_name" LIKE \'%man%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" IN (\'man\') OR "submission_frozenauthor"."last_name" IN (\'man\')) AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:30:48.762090\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21',
'time': '0.002'}, 
{'sql': 'SELECT DISTINCT "submission_article"."id", "submission_article"."journal_id", "submission_article"."owner_id", "submission_article"."title", "submission_article"."subtitle", "submission_article"."abstract", "submission_article"."non_specialist_summary", "submission_article"."language", "submission_article"."section_id", "submission_article"."license_id", "submission_article"."is_remote", "submission_article"."remote_url", "submission_article"."correspondence_author_id", "submission_article"."competing_interests_bool", "submission_article"."competing_interests", "submission_article"."render_galley_id", "submission_article"."date_started", "submission_article"."date_accepted", "submission_article"."date_declined", "submission_article"."date_submitted", "submission_article"."date_published", "submission_article"."date_updated", "submission_article"."current_step", "submission_article"."page_numbers", "submission_article"."stage", "submission_article"."publication_fees", "submission_article"."submission_requirements", "submission_article"."copyright_notice", "submission_article"."comments_editor", "submission_article"."large_image_file_id", "submission_article"."exclude_from_slider", "submission_article"."thumbnail_image_file_id", "submission_article"."peer_reviewed", "submission_article"."is_import", "submission_article"."article_agreement", "submission_article"."ithenticate_id", "submission_article"."ithenticate_score", "submission_article"."primary_issue_id", "submission_article"."meta_image", "submission_article"."is_preprint", "submission_article"."preprint_decision_notification", "submission_article"."preprint_journal_article_id" FROM "submission_article" LEFT OUTER JOIN "submission_article_keywords" ON ("submission_article"."id" = "submission_article_keywords"."article_id") LEFT OUTER JOIN "submission_keyword" ON ("submission_article_keywords"."keyword_id" = "submission_keyword"."id") LEFT OUTER JOIN "submission_frozenauthor" ON ("submission_article"."id" = "submission_frozenauthor"."article_id") WHERE ("submission_article"."is_preprint" = 0 AND ("submission_article"."title" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_keyword"."word" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_article"."subtitle" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_frozenauthor"."last_name" LIKE \'%bur%\' ESCAPE \'\\\' OR "submission_frozenauthor"."first_name" IN (\'bur\') OR "submission_frozenauthor"."last_name" IN (\'bur\')) AND "submission_article"."journal_id" = 1 AND "submission_article"."stage" = \'Published\' AND "submission_article"."date_published" <= \'2019-03-20 15:31:45.396914\') ORDER BY "submission_article"."date_published" DESC, "submission_article"."title" ASC LIMIT 21', 
'time': '0.001'}]
```
total time for all this was 0.027. author-related searches account for .010 of those.

Reply:
There are multiple queries in your output (e.g. queries for the search terms `"rich"` and `"man"`), from what I can see only 2 of them are for the  search in your script and they took 0.001s each, which is also probably rounded up. It also shows that it took 0.009 to query a journal by indexed primary key, which is an outlier in this case as I imagine it is including loading the sqlite connection for the first time.

I've run that script in our DB and it is slow, as it has to do a full scan of 3 tables, but this will be the case even if we don't include the check for the name and surname.

I've gotten better results with this query, which won't reduce the cardinality but simplifies the filtering:
```
author_search = r"|".join(search_term.split(" "))
a = submission_models.Article.objects.filter(
(Q(title__icontains=search_term) |
Q(keywords__word__icontains=search_term) |
Q(subtitle__icontains=search_term)) |
(Q(frozenauthor__first_name__regex=author_search) |
Q(frozenauthor__last_name__regex=author_search)),
journal=j, 
stage=submission_models.STAGE_PUBLISHED,
date_published__lte=timezone.now()
).distinct()
```
By matching with the regex, we avoid the need for the `IN` checks although the time complexity is still O(n)

Reply:
I've been testing (trying to break) this and getting some results that are not really ideal with this author search regex command.
Having just a space will return everything.
Limiting to a minimum of 3 characters from the form and adding `.strip()` before `.split()` handles this okay. But searching something like "p r o" will search for authors with a P an R or O in their name `p|r|o`
So searching for someone's full name who is like "F Scott Fitzgerald" would lookup anyone who has an F.  Limiting these to be also > 3 characters would prevent you from searching a name like  "Miguel de Cervantes" or anyone named "Ng". Perhaps we could concat the first name and last name here and still perform the .split() search.
https://stackoverflow.com/questions/32559106/django-queryset-filter-after-a-concatenation-of-two-columns ?

Reply:
Well, now it is a question on how flexible we want to be in the search:
Should searching for  "SCO" return articles from authors "F Scott Fitzgerald", "Francisco Vega", with the keywords "UNESCO", "SCO"... 
I think perhaps we should take a step back and match author names/last names exactly (other than, maybe, case sensitivity, which is what the original implementation was lacking)

Thoughts @ajrbyers?

Reply:
I think perhaps we’ve overthought the author search. Generally speaking if someone searches by author they are aware of who they are searching for eg. For Mauro I would search “Sanchez” his full name. That said we should consider adding support for external search tools that will be better suited to this task and allow folks to configure this sort of thing.